agent/src/test/groovy/nu/studer/teamcity/buildscan/agent/gradle/BaseInitScriptTest.groovy

```diff
@@ -180,29 +188,45 @@ class BaseInitScriptTest extends Specification {
         }
     }
 
-    BuildResult run(GradleVersion gradleVersion = GradleVersion.current(), List<String> jvmArgs = [], Map<String, String> envVars = [:]) {
-        createRunner(gradleVersion, jvmArgs, envVars).build()
-    }
+    BuildResult run(GradleVersion gradleVersion = GradleVersion.current(), TcPluginConfig tcPluginConfig = new TcPluginConfig(), List<String> additionalJvmArgs = []) {
 
-    BuildResult runAndFail(GradleVersion gradleVersion = GradleVersion.current(), List<String> jvmArgs = [], Map<String, String> envVars = [:]) {
-        createRunner(gradleVersion, jvmArgs, envVars).buildAndFail()
-    }
+        DefaultGradleRunner testKitRunner = new DefaultGradleRunner().withProjectDir(testProjectDir)
+            .withGradleVersion(gradleVersion.version)
+            .forwardOutput() as DefaultGradleRunner
 
-    GradleRunner createRunner(GradleVersion gradleVersion = GradleVersion.current(), List<String> jvmArgs = [], Map<String, String> envVars = [:]) {
-        def args = ['tasks', '-I', initScriptFile.absolutePath]
+        // Provide BuildScanServiceMessageInjector with Gradle User Home for testkit
+        File gradleUserHome = testKitRunner.testKitDirProvider.dir
+        def injector = new TestBuildScanServiceMessageInjector(gradleUserHome, EventDispatcher.create(AgentLifeCycleListener.class), Mock(ExtensionApplicationListener))
 
-        def runner = ((DefaultGradleRunner) GradleRunner.create())
-            .withJvmArguments(jvmArgs)
-            .withGradleVersion(gradleVersion.version)
-            .withProjectDir(testProjectDir)
-            .withArguments(args)
-            .forwardOutput()
+        TestContext context = new TestContext(tcPluginConfig.runner, agentTempDir, tcPluginConfig.toConfigProperties(), [:])
+        injector.beforeRunnerStart(context)
 
-        if (envVars) {
-            runner.withEnvironment(envVars)
+        def args = ['tasks']
+        def gradleArgs = context.runnerParameters.get("ui.gradleRunner.additional.gradle.cmd.params")
+        if (gradleArgs) {
+            args += (gradleArgs.split(' ') as List<String>)
         }
+        testKitRunner.withArguments(args)
 
-        runner
+        def testKitSupportsEnvVars = gradleVersion.baseVersion >= GRADLE_3_5.gradleVersion
+        if (testKitSupportsEnvVars) {
+            testKitRunner.withEnvironment(context.buildParameters.environmentVariables)
+            testKitRunner.withJvmArguments(additionalJvmArgs)
+        } else {
+            testKitRunner.withJvmArguments(tcPluginConfig.toSysProps() + additionalJvmArgs)
```

Comment:
I think we can make this nicer (and less implicit) if instead of using `toSysProps` here, we have a table in this base class that allows us to map the env vars understood by the init script to sys props. We already have all the constants in the injector class, thus this should be easy to pour into a test-scoped lookup map.

Reply:
We can then also remove the limitations documented on the cmd line runner func tests.

Reply:
Fixed in 371e4d19c37724526026d600029097392937be30

Reply:
Great.